p2panda-rs/src/materialiser/graph.rs

```diff
@@ -244,86 +279,137 @@ impl<'a, T: PartialEq + Clone + std::fmt::Debug> Graph<T> {
                         // is either a cycle or missing links.
                         return Err(GraphError::BadlyFormedGraph);
                     }
-                    // Else don't do anything and break out of this loop.
+
+                    // push _last node we visited_ to merged_branch_tips.
+                    graph_data.merged_branch_tips.push(current_node.data());
+
                     // println!(
                     //     "{}: is merge and does not have dependencies met",
                     //     next_node.key()
                     // );
                     break;
                 }
                 // If it wasn't a merge node, push it to the sorted stack and keep walking.
-                sorted.push(next_node);
-                // println!("{}: sorted to position {}", next_node.key(), sorted.len());
+                sorted_nodes.push(next_node);
+                graph_data.sorted.push(next_node.data());
+
+                // If it is a tip, push it to the graph tips list.
+                if next_node.is_tip() {
+                    graph_data.graph_tips.push(next_node.data());
+                }
+
+                // println!("{}: sorted to position {}", next_node.key(), sorted_nodes.len());
                 current_node = next_node;
             }
         }
-        Ok(sorted.iter().map(|node| node.data()).collect())
+        Ok(graph_data)
     }
 
     /// Sort the entire graph, starting from the root node.
-    pub fn sort(&'a self) -> Result<Vec<T>, GraphError> {
+    pub fn sort(&'a self) -> Result<GraphData<T>, GraphError> {
         let root_node = self.root_node();
         self.walk_from(&root_node.key())
     }
 }
 
-impl<'a, T: PartialEq + Clone + std::fmt::Debug> Default for Graph<T> {
+impl<'a, T: PartialEq + Clone + Debug> Default for Graph<T> {
     fn default() -> Self {
         Self::new()
     }
 }
 
 #[cfg(test)]
 mod test {
+    use crate::materialiser::graph::GraphData;
+
     use super::Graph;
 
     #[test]
     fn basics() {
         let mut graph = Graph::new();
-        graph.add_node("a", "Wake Up");
```

Comment:
Whaaaaat, I was just reading this this morning and was like "aww, thats such a nice way to understand this test". Why did you take it away? :smile:

Reply:
Awwww, sorry, I thought maybe I was just having a silly idea and it didn't actually make sense to others. Also, the simpler node letter naming works better for showing the resolved order. But as you liked it, I'll bring it back for it's own special test ;-p

Reply:
It's back ;-p

Reply:
Jaaaa!!